Code Quality changes

Drapersniper opened the pull request:
### Type

- [x] Bugfix
- [ ] Enhancement
- [ ] New feature

### Description of the changes
It was easier making a new PR than reverting changes ...
This is a variant of #2978 but it doesn't have the PyCharm inspections hints, and it doesn't break l18n 👀 
This one also catches a few typos and mispelling that the original did not.

This PR essentially organises imports alphabetically into 3 groups.

Built-ins
3rd Party
Package

It also removes unused variables, convert lamda assignment to def's to PEP8 stops complaining and some other minor tweaks

Drapersniper committed:
Removes `MAX_BALANCE` from bank, user `bank.get_max_balance()` now
`[p]bankset maxbal` can be used to set the maximum bank balance

Signed-off-by: Guy <guyreis96@gmail.com>

Drapersniper committed:
Bring it up to date

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Bring it up to date

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Bring it up to date

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Merge branch 'V3/develop' of https://github.com/Cog-Creators/Red-DiscordBot into V3/develop

Drapersniper committed:
Code health + Typo fixes

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Black

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Black

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
changelogs

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

mikeshardmind commented:
For other people coming across some of this later, as some discussion has happened in discord.

We should avoid using `_` as a placeholder var name anywhere in red's codebase due to interactions with gettext and i18n use.

It's fine to rename something to `_loop_counter` or something else which indicates to tools that the var name isn't important.

Drapersniper committed:
Revert some of changes based on discussion that has happened on Discord
Replace `_` in loops to `_loop_counter`

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

mikeshardmind commented on redbot/cogs/mod/mutes.py:
```diff
@@ -114,7 +116,7 @@ class MuteMixin(MixinMeta):
             )
         except RuntimeError as e:
-            await ctx.send(e)
+            await ctx.send(str(e))
```
This adds an unneeded overhead to each case of it. Better to fix the typhehint on `ctx.send` see #2520 for more details on things needed around this.

mikeshardmind commented on redbot/cogs/mod/names.py:
```diff
@@ -131,10 +132,12 @@ class ModInfo(MixinMeta):
                 # to every single check running on users than the occasional user info invoke
-                # We don't start by building this way, since the number of times we hit this should be
-                # infintesimally small compared to when we don't across all uses of Red.
+                # We don't start by building this way,
```
Why is this PR adjusting whitespace in comments outside of a line-length goal?

Drapersniper replied:
Same here like length was longer than 99, for some reason black didn’t complain for some of these but IDE flagged them as being too long (usually 100/110 characters

mikeshardmind commented on redbot/core/commands/context.py:
```diff
@@ -28,7 +29,7 @@ def __init__(self, **attrs):
 
     async def send(self, content=None, **kwargs):
-        """Sends a message to the destination with the content given.
+        r"""Sends a message to the destination with the content given.
```
Why?

Tobotimus replied:
`\*` down in the docstring below is an invalid escape character, and is technically deprecated since Python 3.6. Sometimes people escape `**kwargs` in docstrings with `\*\*kwargs` (same thing with `*args` -> `\*args`), thinking that sphinx will start bold/emphasis formatting. But sphinx actually handles this, and suggests omitting the backslashes for varargs: https://www.sphinx-doc.org/en/master/usage/extensions/example_numpy.html

So, **TL;DR**, don't use `r"""` docstrings containing `\*\*kwargs`, use plain unicode `"""` docstrings containing `**kwargs`.

mikeshardmind replied:
My "why" was more, why not fix the root issue here, just not expressing it as such.

mikeshardmind commented on redbot/cogs/warnings/warnings.py:
```diff
@@ -361,7 +360,7 @@ def __init__(self, bot: Red):
 
             try:
-                userid: int = user.id
+                user.id
```
No. This should be changed to use `hasattr` rather than try/except if we aren't using `userid` below

mikeshardmind commented on redbot/cogs/audio/audio.py:
```diff
@@ -467,7 +469,8 @@ def _restart_connect(self):
             "This setting is only for bot owners to set a localtracks folder location "
             "if the Lavalink.jar is being ran from outside of the Audio data directory.\n"
-            "In the example below, the full path for 'ParentDirectory' must be passed to this command.\n"
+            "In the example below, the full path for "
```
Why?

Drapersniper replied:
This is due to line length being longer than 99 characters, one bellow it’s the same but the backlash bellow is so that when this its printed to a user it isn’t forced into a new line

mikeshardmind replied:
Didn't notice the line length on this one was that long, mildly surprised black didn't complain about it before given that it was already inside parens.

mikeshardmind commented on redbot/cogs/audio/audio.py:
```diff
@@ -570,7 +575,8 @@ def _restart_connect(self):
 
         When toggled off, users will be able to play songs from non-commercial websites and links.
-        When toggled on, users are restricted to YouTube, SoundCloud, Mixer, Vimeo, Twitch, and Bandcamp links."""
+        When toggled on, users are restricted to YouTube, SoundCloud, \
```
why? (and also, dont use `\` line continuations here)

Tobotimus replied:
Yeah, this'll get rendered like this in the help command:
```
When toggled on, users are restricted to YouTube, SoundCloud,         Mixer, Vimeo, Twitch, and Bandcamp links.
```

I always get slightly annoyed with how docstrings end up looking in the help command - sticking to the 99-char line length but also considering where the line breaks will appear in the help embed is frustrating.

My solution would be to rewrap/unwrap each paragraph in the help formatter. Code-block help formatter should rewrap, whilst embed help formatter should unwrap. This means a single newline would usually treated as a space (we can make exceptions, possibly for numbered/unordered lists) and double newlines are preserved. Still not sure about this though.

mikeshardmind replied:
Let's keep the docstring free of line continuations here.

I'll open an issue + handle this case for 3.2 in how we handle help.

mikeshardmind reviewed (changes requested):
Some of these changes aren't ready to go.

I'm still mixed on accepting the import sorting without also making it a requirement going forward.
Doing this as is just increases the size of diffs every so often as we resort.

There's a lot of whitespace changes here which don't seem like positives.

Still seems like too much of this is being done to satisfy tools we haven't and probably won't standardize on.

Drapersniper committed:
Update to reflect Sinbad's review

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Better import sorting (isort with custom settings for clarity)

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
caught a few extra typos

Signed-off-by: guyre <27962761+drapersniper@users.noreply.github.com>

Drapersniper commented:
Just for clarity, this should be fully up to date with the review from Sinbad and Toby.

mikeshardmind reviewed (changes requested):
I'm noticing some added `# isort: ignore` statements here.

For this to be included like this, if we're going to sort imports, it really should be added to both the makefile recipe (for reformatting) and the style test. Otherwise, these comments have no place in our code base.

I've already stated my dislike of mixing isort usage with black in discord, but if you have a set of settings and automatable steps which will produce consistent results, that shouldn't be a blocker to this.

I am going to insist that this be automated to keep the diff sizes down rather than something that just gets changed every so often.

Drapersniper commented:
> I'm noticing some added `# isort: ignore` statements here.
> 
> For this to be included like this, if we're going to sort imports, it really should be added to both the makefile recipe (for reformatting) and the style test. Otherwise, these comments have no place in our code base.
> 
> I've already stated my dislike of mixing isort usage with black in discord, but if you have a set of settings and automatable steps which will produce consistent results, that shouldn't be a blocker to this.
> 
> I am going to insist that this be automated to keep the diff sizes down rather than something that just gets changed every so often.

ermm ... this was unintentional ... i missed that the isort stuff was on this branch, let me remove them for now, there's 1 PR pending on the isort repo that after it's merged we should be able to fully automate to where black and isort can be used together consistently

Drapersniper committed:
Precommit hooks + isort + black

Drapersniper committed:
Merge branch 'isort_black' of https://github.com/Drapersniper/Red-DiscordBot into codehealth

 Conflicts:
	docs/conf.py
	docs/framework_apikeys.rst
	redbot/__init__.py
	redbot/__main__.py
	redbot/cogs/admin/__init__.py
	redbot/cogs/admin/admin.py
	redbot/cogs/admin/announcer.py
	redbot/cogs/admin/converters.py
	redbot/cogs/alias/__init__.py
	redbot/cogs/alias/alias.py
	redbot/cogs/alias/alias_entry.py
	redbot/cogs/audio/__init__.py
	redbot/cogs/audio/audio.py
	redbot/cogs/audio/equalizer.py
	redbot/cogs/audio/errors.py
	redbot/cogs/audio/manager.py
	redbot/cogs/bank/__init__.py
	redbot/cogs/bank/bank.py
	redbot/cogs/cleanup/__init__.py
	redbot/cogs/cleanup/cleanup.py
	redbot/cogs/cleanup/converters.py
	redbot/cogs/customcom/__init__.py
	redbot/cogs/customcom/customcom.py
	redbot/cogs/downloader/__init__.py
	redbot/cogs/downloader/checks.py
	redbot/cogs/downloader/converters.py
	redbot/cogs/downloader/downloader.py
	redbot/cogs/downloader/installable.py
	redbot/cogs/downloader/json_mixins.py
	redbot/cogs/downloader/log.py
	redbot/cogs/downloader/repo_manager.py
	redbot/cogs/economy/__init__.py
	redbot/cogs/economy/economy.py
	redbot/cogs/filter/__init__.py
	redbot/cogs/filter/filter.py
	redbot/cogs/general/__init__.py
	redbot/cogs/general/general.py
	redbot/cogs/image/__init__.py
	redbot/cogs/image/image.py
	redbot/cogs/mod/__init__.py
	redbot/cogs/mod/abc.py
	redbot/cogs/mod/converters.py
	redbot/cogs/mod/events.py
	redbot/cogs/mod/kickban.py
	redbot/cogs/mod/mod.py
	redbot/cogs/mod/movetocore.py
	redbot/cogs/mod/mutes.py
	redbot/cogs/mod/names.py
	redbot/cogs/mod/settings.py
	redbot/cogs/mod/slowmode.py
	redbot/cogs/modlog/__init__.py
	redbot/cogs/modlog/modlog.py
	redbot/cogs/permissions/__init__.py
	redbot/cogs/permissions/converters.py
	redbot/cogs/permissions/permissions.py
	redbot/cogs/reports/__init__.py
	redbot/cogs/reports/reports.py
	redbot/cogs/streams/__init__.py
	redbot/cogs/streams/streams.py
	redbot/cogs/streams/streamtypes.py
	redbot/cogs/warnings/__init__.py
	redbot/cogs/warnings/helpers.py
	redbot/cogs/warnings/warnings.py
	redbot/core/__init__.py
	redbot/core/bank.py
	redbot/core/bot.py
	redbot/core/checks.py
	redbot/core/cli.py
	redbot/core/cog_manager.py
	redbot/core/commands/__init__.py
	redbot/core/commands/commands.py
	redbot/core/commands/context.py
	redbot/core/commands/converter.py
	redbot/core/commands/help.py
	redbot/core/config.py
	redbot/core/core_commands.py
	redbot/core/data_manager.py
	redbot/core/dev_commands.py
	redbot/core/drivers/__init__.py
	redbot/core/drivers/base.py
	redbot/core/drivers/json.py
	redbot/core/drivers/log.py
	redbot/core/drivers/mongo.py
	redbot/core/drivers/postgres/__init__.py
	redbot/core/drivers/postgres/postgres.py
	redbot/core/errors.py
	redbot/core/events.py
	redbot/core/i18n.py
	redbot/core/modlog.py
	redbot/core/rpc.py
	redbot/core/utils/__init__.py
	redbot/core/utils/antispam.py
	redbot/core/utils/caching.py
	redbot/core/utils/chat_formatting.py
	redbot/core/utils/common_filters.py
	redbot/core/utils/embed.py
	redbot/core/utils/mod.py
	redbot/core/utils/predicates.py
	redbot/core/utils/safety.py
	redbot/core/utils/tunnel.py
	redbot/launcher.py
	redbot/logging.py
	redbot/pytest/__init__.py
	redbot/pytest/admin.py
	redbot/pytest/alias.py
	redbot/pytest/cog_manager.py
	redbot/pytest/core.py
	redbot/pytest/data_manager.py
	redbot/pytest/downloader.py
	redbot/pytest/economy.py
	redbot/pytest/mod.py
	redbot/pytest/permissions.py
	redbot/pytest/rpc.py
	redbot/setup.py
	setup.py
	tests/cogs/admin/test_admin.py
	tests/cogs/downloader/test_downloader.py
	tests/cogs/downloader/test_installable.py
	tests/cogs/test_alias.py
	tests/cogs/test_economy.py
	tests/cogs/test_mod.py
	tests/cogs/test_permissions.py
	tests/cogs/test_trivia.py
	tests/conftest.py
	tests/core/test_cog_manager.py
	tests/core/test_commands.py
	tests/core/test_config.py
	tests/core/test_data_manager.py
	tests/core/test_installation.py
	tests/core/test_rpc.py
	tests/core/test_utils.py
	tests/core/test_version.py

Drapersniper committed:
Test to run against 3.7.5

Drapersniper committed:
Mention iSort is also being run

Drapersniper committed:
Fix tests

Drapersniper committed:
Missed 1 towncrier entry

Drapersniper committed:
For some reason the pre-commit hook for black was behaving differently ... so lets get rid of that file

Drapersniper committed:
Typos and more Typos
also
Fixed some unbound variables
Fixed some unused variables
Fixed some redundant escape characters
Fixed some redundant parenthesis
Made some methods stattic

Drapersniper committed:
Replace _ to _loop_counter across the project (When not used for translation)

Drapersniper committed:
*sigh*

Drapersniper committed:
Missed some capitalization

Drapersniper commented:
Okay so Implemented both iSort and Black into tests, this is using iSort settings reccomended by the official Black docs.

All test are passing and no compatibility issues.

I would say if we want go ahead and use iSort in the project for further formatting consistency this PR is ready and could be merged after 3.2 (to reduce the number of conflicts just priort to a major release)

Drapersniper committed:
Add working precommit-hook config

Signed-off-by: Drapersniper <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Merge branch 'V3/develop' of https://github.com/Cog-Creators/Red-DiscordBot into codehealth

Kowlin commented on .travis.yml:
```diff
@@ -30,10 +30,10 @@ jobs:
     - stage: PyPi Deployment
       if: tag IS present
-      python: 3.7.2
+      python: 3.7.5
```
Lets not change deployment versions... they are currently working on black magic and I don't want anything breaking with them.

Drapersniper replied:
WUT ... I though I updated these to 3.7.5 .... *sigh* 
noted i'll ensure its 3.7.5 and py37

Drapersniper replied:
oh wait im dumb .... shouldn't at least the TOXENV be py37, since the bot no longer supports python3.6 so running the test agaisnt 3.7 would be more realist and find bugs caused by an specific version of python?

Kowlin commented on .travis.yml:
```diff
@@ -46,10 +46,10 @@ jobs:
     - stage: Crowdin Deployment
       if: tag IS present
-      python: 3.7.2
+      python: 3.7.5
```
Same over here.

Kowlin commented on .travis.yml:
```diff
@@ -30,10 +30,10 @@ jobs:
       env:
         - DEPLOYING=true
-        - TOXENV=py36
+        - TOXENV=py37
```
Same over here.

Kowlin commented on .travis.yml:
```diff
@@ -46,10 +46,10 @@ jobs:
       env:
         - DEPLOYING=true
-        - TOXENV=py36
+        - TOXENV=py37
```
Same over here.

mikeshardmind commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,36 @@
+default_language_version:
```
Adding this pre-commit config like this is a strong No from me. It requires we maintain settings in multiple places. We have configurations for these already, we have tests that will fail if someone didn't comply, and we have an easy way to run anything which should be automated. IDEs will format with black using our settings already just by setting black as the formatting provider. pre-commit runs in an isolated environment each time, ignoring our central settings. 

As an example of the trouble this can cause, this has args which mismatch the ones we are using for one of the tools here, and you just created it. The fewer places we duplicate settings, the better.

This just isn't suited to the project.

Drapersniper replied:
Are we sure thats the case when I was running the pre-commit hook localy based on this file it was using the settings from the setup.cfg, the individual files and the .ini file 
Reason I added here was something user could easily run to process all the formatting  in 1 command, and ensure everything will pass the Style check while applying the extra style stuff thats black and isort compatible thats in the pre-commit hooks (like json formatting etc)

mikeshardmind replied:
`make reformat` already handles the style bit. And yes, pre-commit creates a virtualenv specifically for all of this, using the versions specified here, not in setup.cfg and using the args specified here over the ones in our configuration files, which makes this another place we would need to update multiple things rather than just let people be responsible for their own development environments and catch it at the tests.

mikeshardmind commented on .isort.cfg:
```diff
@@ -0,0 +1,19 @@
+force_grid_wrap = 0
+sections = FUTURE,STDLIB,THIRDPARTY,FIRSTPARTY,LOCALFOLDER
+default_section = THIRDPARTY
+import_heading_stdlib = Standard Library
```
All of these headings are just visual noise. If isort is being enforced, they don't do anything useful, isort will sort it out anyway.

Drapersniper replied:
So yeah this came up on discord before, yes they are noise but ... it helps people who are less familiar with the ordering  prior to running isort. So this is more educational than anything happy to remove but  from the discord talks they would be here to explain the ordering so people  learn  it rather than just having a tools "randomly" sort things priort to commiting.

mikeshardmind replied:
Info like this belongs in contributing.md, not in every source code file. This isn't a critical comment that needs visibility everywhere something is in use.

mikeshardmind reviewed (changes requested):
I haven't reviewed this in full. I'm still skeptical of a lot of the changes here at a surface level.

I had a few specific comments, but also I don't think it's appropriate here to be adding 
`# -*- coding: utf-8 -*-` to the top of each python file.

This is a python3.7+ project. Every python file is treated as utf-8 by default as of python3.
Encoding specifications are only neccessary when they differe from the expected, It's just noise here and doesn't even draw contrast in any files where we might need a different encoding because all of the files are utf-8.

Drapersniper commented:
> I haven't reviewed this in full. I'm still skeptical of a lot of the changes here at a surface level.
> 
> I had a few specific comments, but also I don't think it's appropriate here to be adding
> `# -*- coding: utf-8 -*-` to the top of each python file.
> 
> This is a python3.7+ project. Every python file is treated as utf-8 by default as of python3.
> Encoding specifications are only neccessary when they differe from the expected, It's just noise here and doesn't even draw contrast in any files where we might need a different encoding because all of the files are utf-8.

So this is the case for people with a real IDE, but in the past when using a non python IDE i did hit issues when editing on Windows  and handling emojis  and the emojis breaking

Kowlin commented:
That's the fault of your IDE and not the project. Visual studio code on Windows has fine emoji support.

Drapersniper committed:
address reviews

Signed-off-by: Drapersniper <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
revert a this stupid comma + address a few other things i missed

Signed-off-by: Drapersniper <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Remove unused imports without breaking shit :aha:

Signed-off-by: Drapersniper <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Lets not break the docs

Signed-off-by: Drapersniper <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Remove redundant parenthesis and update some typehints

mikeshardmind commented:
Marking this is blocked for 2 reasons.

1. Let's wait on isort 5.0.0, it's supposed to be getting some extra stuff to make it cooperate with black a bit better.

2. This still appears to be touching files that were generated, not written (localization messages) which I'm sure isn't in our interests to have multiple tools changing.

Drapersniper committed:
revert - locale changes

Signed-off-by: Drapersniper <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Merge branch 'V3/develop' of https://github.com/Cog-Creators/Red-DiscordBot into codehealth

Drapersniper committed:
Revert locale file changes

Signed-off-by: Drapersniper <27962761+drapersniper@users.noreply.github.com>
